schema/array.go

```diff
@@ -26,21 +26,51 @@ func (v *Array) Compile(rc ReferenceChecker) (err error) {
 	return
 }
 
-// Validate implements FieldValidator.
-func (v Array) Validate(value interface{}) (interface{}, error) {
+func (v Array) validate(value interface{}, query bool) ([]interface{}, error) {
 	arr, ok := value.([]interface{})
 	if !ok {
 		return nil, errors.New("not an array")
 	}
 	for i, val := range arr {
 		if v.Values.Validator != nil {
-			val, err := v.Values.Validator.Validate(val)
+			var err error
+			if query {
+				if queryValidator, ok := v.Values.Validator.(FieldQueryValidator); ok {
+					val, err = queryValidator.ValidateQuery(val)
+				} else {
+					val, err = v.Values.Validator.Validate(val)
+				}
+			} else {
+				val, err = v.Values.Validator.Validate(val)
+			}
 			if err != nil {
 				return nil, fmt.Errorf("invalid value at #%d: %s", i+1, err)
 			}
 			arr[i] = val
 		}
 	}
+	return arr, nil
+}
+
+// ValidateQuery implements FieldQueryValidator.
+func (v Array) ValidateQuery(value interface{}) (interface{}, error) {
+	val, err := v.validate(value, true)
+	if err != nil {
+		return nil, err
+	}
+	// if there is no valid items in the array, then return nil, instead of []
+	if len(val) == 0 {
+		return nil, nil
+	}
+	return val, nil
+}
+
+// Validate implements FieldValidator.
+func (v Array) Validate(value interface{}) (interface{}, error) {
+	arr, err := v.validate(value, false)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Performance wise, it seams more sane to do the lengh check first. That probably means doing the type assertion to an array before calling `validateValues`.